File: Python/CNN/CNN.py
import pandas as pd
import numpy as np
from sklearn.preprocessing import MinMaxScaler

# The 2 following lines may show errors on import, but the venv is ignoring the errors
from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Conv1D, MaxPooling1D, Flatten, Dense

import matplotlib.pyplot as plt

# Loading the dataset
file_path = "../../Dataset/CNN/airPassengers.csv"
df = pd.read_csv(file_path)

# Defining the time index
df["month"] = pd.to_datetime(df["month"])
df.set_index("month", inplace=True)

# Normalizing the data
scaler = MinMaxScaler()
df_scaled = scaler.fit_transform(df)


# Function to create sequences for time series data training
def create_sequences(data, seq_length):
    sequences = []
    targets = []
    data_len = len(data)

    for i in range(data_len - seq_length):
        seq = data[i:i+seq_length]
        target = data[i+seq_length]
        sequences.append(seq)
        targets.append(target)

    return np.array(sequences), np.array(targets)


# Defining tge sequence length
sequence_length = 12

# Creating sequences and targets
X, y = create_sequences(df_scaled, sequence_length)

# Splitting the data into training and testing sets
split_ratio = 0.8
split_index = int(split_ratio * len(X))

X_train, X_test = X[:split_index], X[split_index:]
y_train, y_test = y[:split_index], y[split_index:]

# Reshaping the data for CNN input
X_train = X_train.reshape((X_train.shape[0], X_train.shape[1], 1))
X_test = X_test.reshape((X_test.shape[0], X_test.shape[1], 1))

# Building the CNN model
model = Sequential()
model.add(Conv1D(filters=64, kernel_size=2, activation="relu", input_shape=(sequence_length, 1)))
model.add(MaxPooling1D(pool_size=2))
model.add(Flatten())
model.add(Dense(50, activation="relu"))
model.add(Dense(1))

# Compiling the model
model.compile(optimizer="adam", loss="mse")

# Training the model
model.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_test, y_test))

# Evaluating the model
loss = model.evaluate(X_test, y_test)
print(f"Mean Squared Error on Test Data: {loss}")

# Making predictions
predictions = model.predict(X_test)

# Inverse transform the predictions and true values to the original scale
predictions_inv = scaler.inverse_transform(predictions)
y_test_inv = scaler.inverse_transform(y_test.reshape(-1, 1))

# Plotting the results
plt.figure(figsize=(12, 6))
plt.plot(df.index[-len(y_test):], y_test_inv, label="True Values")
plt.plot(df.index[-len(predictions_inv):], predictions_inv, label="Predictions")
plt.title("Air Passengers Time Series Prediction with CNN")
plt.xlabel("Year")
plt.ylabel("Number of Passengers")
plt.legend()
plt.show()

File: Julia/CNN/CNN.jl
using CSV, DataFrames, Dates, Flux, Statistics, MLUtils

# Loading CSV file
df = CSV.File("Dataset/CNN/airPassengers.csv") |> DataFrame

# Convert 'month' column in Date type
df.month = Dates.Date.(string.(df.month), "yyyy-mm-dd")

# Extract year and month as numerical features
df.year = Dates.year.(df.month)
df.month = Dates.month.(df.month)

# Replace months by a sequence of numbers from 1 to the length of the months column
df.month = 1:length(df.month)

# Extracting time sequences
sequences = [Float32.(df.passengers)]

# Splitting training data and test data
train_data, test_data = Flux.splitobs(sequences[1], at=0.8)

# Defining the time "window" for the CNN 1D
window_size = 12

# Preparing the time series data
X = []
y = []
for i in 1:length(train_data) - window_size
    push!(X, train_data[i:i+window_size-1])
    push!(y, train_data[i+window_size])
end

# Reshaping X and y to add a dimension
X = reduce(hcat, X)'
X = reshape(X, (:,window_size, 1))
y = reshape(y, (length(y), 1))

# Creating the dataloader for (X,y)
data_loader = DataLoader((X, y), batchsize=10, shuffle = true)

# Defining the architecture for the model
input_size = (window_size, 1)
num_classes = 1

# Defining the 1D CNN architecture
function create_cnn(input_size, num_classes)
    return Chain(
        Conv((2, 1), 1=>16, relu),
        MaxPool((2, 1)),
        Dense(16, num_classes),
    )
end

cnn_model = create_cnn(input_size, num_classes)

# Defining loss function and the optimizer
loss(x, y) = Flux.mse(cnn_model(x), y)
optimizer = ADAM(0.001)

# Training the model
epochs = 100
for epoch in 1:epochs
    for (X, y) in data_loader
        X = cat(X..., dims=2)
        y = cat(y..., dims=2)
        Flux.train!(loss, Flux.params(cnn_model), [(X, y)], optimizer)
    end
    println("Epoch: $epoch, Loss: $(loss(X, y))")
    println("Model weights: ", [params(cnn_model)[i] for i in 1:length(params(cnn_model))])
    println("Loss on entire training data: $(loss(X, y))")
end

# Evaluate the model
test_X = cat(test_data..., dims=2)[1:end-window_size, :]
test_y = cat(test_data..., dims=2)[window_size+1:end, :]

mse = Flux.mse(cnn_model(test_X), test_y)

println("Mean Squared Error on Test Data: $mse")